### ISSUE TITLE
Include project uuid in the links.json file in the bundle

### ISSUE THREAD

**Issue by aaclan-ebi** [HumanCellAtlas/ingest-central#453]
User story: As a bundle consumer, I want to see the project uuid in the links.json file, so that I can easily query all of the data related to a project

Requested from slack: see https://humancellatlas.slack.com/archives/C49TW3RV5/p1562091820050600

**Comment by aaclan-ebi** [comment]
my estimate: 2

**Comment by MDunitz** [comment]
This change will allow query service to more easily connect all of the files in a single project which will make that information more available to users.

**Comment by justincc** [comment]
We need to consider actually describing `links.json`

**Comment by justincc** [comment]
See #458 for the documentation ticket

**Comment by aaclan-ebi** [comment]
links.json follows a schema: https://schema.dev.data.humancellatlas.org/system/1.1.5/links
sample links.json: https://dss.dev.data.humancellatlas.org/v1/files/73bd9dd3-13bf-4cbd-914a-ae4a79a7629b/?replica=aws

**Comment by justincc** [comment]
Good point. I feel we need a bit more info though on the fact this file even exists and why.

**Comment by rolando-ebi** [comment]
@MDunitz What should project be linked to? If I understand correctly, project.json should be present in all bundles. All other files are then implicitly related to the project.

**Comment by lauraclarke** [comment]
@rolando-ebi would it be reasonable to say that all bundles only contain one donor? while it isn't strictly speaking correct, I wonder if linking the project to the root of the graph, the donor, might work best

**Comment by rolando-ebi** [comment]
@lauraclarke They always do from what we've seen so far but it's possible for users to submit a donor-less experiment/dataset. But yeah, we could choose to link it to the root of the graph(the root of the graph will always be present) rather than hardcode the link to a donor, which will have the same desired effect.

**Comment by lauraclarke** [comment]
Theh root of the graph sounds like a good solution,

**Comment by MDunitz** [comment]
I think the root of the graph is fine or separate from the graph at the same level as the links array if that wont break anything else. Because there isnt a protocol for deriving the donors from the project so it wouldnt follow the same pattern as the rest of the graph

**Comment by diekhans** [comment]
Given that multiple versions of metadata instances are on their way, it would be good to record FQIDs rather than raw UUIDs.  conversion from FQID to UUiD is trivial, conversion from UUID to FQID requires a lookup in the bundle manifest.

**Comment by tburdett** [comment]
When a user submits any metadata for a project, the project is linked to every piece of metadata in that submission (IIRC). These links don't have explicit semantics (JSON-LD anyone?), but you can consider it to represent `biomaterial|file|process|protocol -- contained_by --> project`. 

Here's my quick cost/benefit analysis. On export to the datastore, I would see the options as one of:

1. Serialize *only* links to the root of the graph. This has the benefit of only adding a small number of additional links into links.json.
1. Serialise *all* of these links between projects and other metadata entities into links.json. This has the benefit of requiring no domain-specific logic in the ingest exporter (i.e. if `projects` rel is present in the ingest API, add the target UUID to links.json).
1. Modify the domain representation in ingest to make it clear that projects only `contain` the entity at the root of the graph. This has the benefit of linking many fewer entities, whilst ensuring the knowledge being submitted is consistent with the information presented by the datastore.

There is the possibility that 1. introduces potentially complex domain-specific logic into the exporter. What happens when a donor is reused across projects (i.e. root of the graph links to two projects, data files only come from one)? How about when a cell line is generated from a donor in one project and reused in another project (links to two projects at different locations in the graph)? What happens when data files are submitted first, with biomaterial metadata added later (root of the graph changes)? The exporter will have to make decisions about these scenarios. Also, given @MDunitz's stated aim of:

> connect all of the files in a single project

...DCP components will have to create workarounds to reverse engineer this set of decisions downstream (e.g. query service/browser has to walk the graph to infer the "contains" relationship to files). That will demand documentation which may be quite complex.

I don't think I understand the downside of 2. It definitely adds many links into a links.json but that doesn't create a high storage cost. Is it costly to generate?

Finally, I assume that 3. doesn't accurately represent what data consumers will expect ("which project was this data file generated by?"). It seem to me likely that submitters will be confused ("what should I link to a project?") and it results in the same downstream reverse engineering scenario as 1.

**Comment by justincc** [comment]
I was hoping we could just stick a `"project-uuid":"b76ecd4d-1701-4795-92f1-dcf274d21366"` key-value pair at the root of the JSON rather than having to explicitly link it to something else (or everything else). Is that the minimum you need @MDunitz ?

If we can't do this then we need to properly thrash out the pros and cons. `links.json` is forming part of our API if it's used by query service - we already suspect analysis are using it (@samanehsan @rexwangcc any comment?). It looks like explicitly linking in project UUID makes this a data model question. Hence #458 to find a place to document this properly.

**Comment by tburdett** [comment]
@justincc the `"project-uuid"` solution is similar to my 1) above, I think, and it comes with many of the same downsides e.g. when there are multiple projects in a a bundle, clients need logic to infer what that means. Completely agreed with the importance is capturing the rationale for `links.json`.

**Comment by diekhans** [comment]
Conceptually, I view project as a container for data/metadata, not part of the experimental graph.  Project will have a many to many relationship to metadata instances.  There are already plans for samples to be shared between projects.

**Comment by MDunitz** [comment]
I agree that projects are a container rather than a part of the graph. However I think we need a way to make it clear that some parts of the graph can belong to multiple projects and make it simple to pull the data relevant to a single project. I think that adding the project id as Justin suggests would be ideal.  Re Tony's concerns
```
What happens when a donor is reused across projects (i.e. root of the graph links to two projects, data files only come from one)?
```
I would assume in this situation a different protocol would be applied and/or there would be different output files, creating a different process to separate the two projects/show where the graphs branch apart. 
If multiple projects are sharing processes, but not the entire graph I think that makes it more important to specify what project(s) it belongs to. 
```
 How about when a cell line is generated from a donor in one project and reused in another project (links to two projects at different locations in the graph)? 
```
In this case I think we should show the creation of the cell line as belonging to both projects
```
What happens when data files are submitted first, with biomaterial metadata added later (root of the graph changes)? The exporter will have to make decisions about these scenarios.:

``` 
this shouldn't be an issue (from query service's perspective)

Also I was under the impression that a bundle would never contain files from multiple projects. 
Some of the data contained in a bundle might be reused in another project but I think that should be represented by pointing to that file in a bundle representing that project.

**Comment by tburdett** [comment]
> a bundle would never contain files from multiple projects

Data files will almost certainly only ever belong to one project. However, it is very likely that data will soon be submitted, though, where biological materials used to generate this data comes from a different project to the data file.

So: you _could_ just add a `project_uuid` to the top of `links.json` that contains the project UUID that contains the data files in this bundle. It's _probably_ safe to use that there will only be one of those (although I think this fails for analysis bundles, as analysis files aren't considered to be contained by any project currently). This still has the result of creating an inconsistent representation of the graph for submitters and consumers.

A whiteboard would help here :)

**Comment by diekhans** [comment]
I believe there will eventually be projects composed of data from other projects.  If copy forward is preserved, then a bundle will have multiple projects.  If we drop copy forward,. then it would be possible to have a restriction of one project per bundle.

Thinking of projects as a many to many SQL relationships seems a natural way to think about it.

**Comment by justincc** [comment]
How about this line of reasoning?

1) A bundle is produced for a project.
2) At the moment bundles contain files from a single project.
3) When projects reference other projects, we remove copy-forward and replace it with an FQID-based referencing scheme.

In this case, we can stick `project-uuid` in the root of `links.json`.

**Comment by lauraclarke** [comment]
Having chatted a bit with @tburdett about this I think his proposal 2 is the right solution

> Serialise all of these links between projects and other metadata entities into links.json. This has the benefit of requiring no domain-specific logic in the ingest exporter (i.e. if projects rel is present in the ingest API, add the target UUID to links.json).

This means no consumer needs any knowledge of what the project id at the top of the links.json actually means. 

It also means some entities in links.json can be linked to more than one project that is fine as if someone needs to ask a specific question about specific other entities in the graph such as Madison's what projects do these files belong to, they can do that without any additional logic

Do not that one donor in multiple projects as a requirement is on the near horizon as all the MRC development atlas projects will share donor fetuses

**Comment by justincc** [comment]
Can @tburdett give an example of what such a `links.json` looks like? It's not clear to me what this means when it's expressed in low-level ingest implementation details.

**Comment by tburdett** [comment]
I don't have an implementation in mind - and I don't want to propose any particular solution. I have my view but it isn't very valid, as I won't actually have to implement it...

But to answer the question, some options I could think of off the top of my head would be:

1) To have project links represented like other links in links.json with inputs/outputs but no processes/protocols, like...

```
{
	...
	<snip>
	...
	"inputs": [
		"ec5cd5c4-54b9-4c1f-894a-e918273786fb"
	],
	"input_type": "biomaterial",
	"outputs": [
		"82ea633d-fd7e-4221-b493-4f21c8c689b2"
	],
	"output_type": "project",
}
```

2) To add an extra block to links.json serializing the project links. Something like...

```
{
	...
	<snip>
	...
	"project_links": [{
			"inputs": [
				"ec5cd5c4-54b9-4c1f-894a-e918273786fb"
			],
			"input_type": "biomaterial",
			"outputs": [
				"82ea633d-fd7e-4221-b493-4f21c8c689b2",
				"be014a2d-260b-4845-8000-22ab41f7b700"
			],
			"output_type": "project",
		}
	]
}
```

3) Add an entirely different type of metadata file to the bundle that simply encodes the project links.

It would be easier to do this with a description of the  usecases for interpreting the metadata graph. I appreciate the desire to be agile here and iterate, but past experience has shown that whenever we require components to agree implicit contracts between them, these decisions get baked in and the "right" set of assumptions to make gets quickly lost.

**Comment by tburdett** [comment]
In order to think any more about this, I think I need to understand: what is the value provided by modifying `links.json` to add a `project-uuid` field, over simply doing a lookup for the project in the bundle you already have?

**Comment by MDunitz** [comment]
The lookup itself is fairly straightforward, I was planning on adding that to handle links files created prior to this update.
 I was told if this was something I wanted/felt would be useful it might be good to contain this information in one place so other people could more easily utilize it. 

Also in talks with @mallory during the f2f we determined that everything in the bundle other than the project (and supplementary files, but they are linked to in the protocol file they are supplementing) is listed in the links.json so including this could be useful if we wanted to move away from bundles

**Comment by justincc** [comment]
Now that this has got stuck in the tarpit of 'the right thing to do' it will almost certainly not happen anytime soon. @MDunitz if you really need it please shout and I will make a command decision even if that creates something we have to deprecate later.

**Comment by diekhans** [comment]
Representing projects it as a container in links.json should be straight-forward.  A list of project ids with the associated entity instance ids.

If there isn't an urgent need for this, I would suggest we hold off until we can actually document the serialization algorithm.
